File: sem3/map/ToyLanguageInterpreter/src/view/textview/TextMenu.java
package view.textview;

import utils.exceptions.ExitTextMenuException;
import utils.exceptions.InterpreterException;
import view.textview.commands.Command;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Scanner;

/**
 * Created by mirko on 14/11/2016.
 */
public class TextMenu {
    private Map<String, Command> commands;
    private Scanner scanner;

    public TextMenu(Scanner scanner) {
        this.commands = new LinkedHashMap<>();
        this.scanner = scanner;
    }

    public void addCommand(Command command) {
        commands.put(command.getKey(), command);
    }

    private void printMenu() {
        for (Command command : commands.values()) {
            String line = String.format("%4s : %s", command.getKey(), command.getDescription());
            System.out.println(line);
        }
    }

    public void show() {
        while (true) {
            printMenu();

            System.out.printf("Type an option: ");
            String key = scanner.nextLine();
            Command command = commands.get(key);

            if (command == null) {
                System.out.println("Option doesn't exist.");
                continue;
            }

            try {
                command.execute(scanner);
            } catch (ExitTextMenuException err) {
                break;
            } catch (InterpreterException err) {
                System.out.println(err);
            }
        }
    }
}

File: sem3/map/ToySharpLanguageInterpreter/ToySharpLanguageInterpreter/src/view/TextMenu.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace ToySharpLanguageInterpreter
{
	public class TextMenu
	{
		private Dictionary<string, Command> commands;

		public TextMenu()
		{
			commands = new Dictionary<string, Command>();
		}

		public void addCommand(Command command)
		{
			commands.Add(command.getKey(), command);
		}

		private void printMenu()
		{
			foreach (KeyValuePair<string, Command> command in commands)
			{
				string line = string.Format("{0} : {1}", command.Value.getKey(), command.Value.getDescription());
				Console.WriteLine(line);
			}
		}

		public void show()
		{
			while (true)
			{
				printMenu();

				Console.Write("Type an option: ");
				string key = Console.ReadLine();
				Command command = commands[key];

				if (command == null)
				{
					Console.WriteLine("Option doesn't exist.");
					continue;
				}

				try
				{
					command.execute();
				}
				catch (ExitTextMenuException)
				{
					break;
				}
				catch (InterpreterException err)
				{
					Console.WriteLine(err.Message);
				}
			}
		}
	}
}